kura/examples/org.eclipse.kura.example.wire.math.singleport.provider/src/main/java/org/eclipse/kura/example/wire/math/singleport/RunningMedian.java

```diff
@@ -0,0 +1,35 @@
+package org.eclipse.kura.example.wire.math.singleport;
+
+import java.util.ArrayDeque;
+import java.util.Arrays;
+
+public class RunningMedian {
+
+    private final ArrayDeque<Double> window;
+    private final int windowSize;
+
+    public RunningMedian(final int windowSize) {
+        this.window = new ArrayDeque<>(windowSize);
+        this.windowSize = windowSize;
+    }
+
+    public double updateAndGetMedian(double value) {
+        int index = 0;
+        if (this.window.size() >= windowSize) {
+            this.window.removeFirst();
+        }
+        this.window.addLast(value);
+        double[] values = new double[windowSize];
+        for (Double windowValue : window) {
```

Comment:
If the size of the queue is large, it would be more efficient to maintain a sorted collection field like a `TreeSet` as the cost of insertion and removal is log(n) instead of n log(n) of `Arrays.sort`. 

In general, `Collection`S provide `toArray` methods to avoid copying the element yourself
Use https://docs.oracle.com/javase/8/docs/api/java/util/ArrayDeque.html#toArray(T[])

Reply:
Set cannot be used because for the nature of it, you cannot store multiple entries with the same value.

Reply:
It can be done with a TreeSet, provided the element contains the value and a monotonic sequence number attached to the value. This would allow for elements with the same value (but a different sequence number).

The complexity of this solution is O(n) as, to get the "central" elements in the TreeSet requires using an iterator (insertion and removal have complexity O(log n))